File: advall-haskell/d13.hs
import Numeric (showIntAtBase)
import Data.Char
import Data.Set (Set, member, union, fromList, size)

isOpen :: (Integer, (Integer, Integer)) -> Integer -> Bool
isOpen (_,(x,y)) luckyNumber = ((length ((filter (== '1')(showIntAtBase 2 
    intToDigit (x*x + 3*x + 2*x*y + y + y*y + luckyNumber) "")))) `mod` 2) == 0

myNotMember :: (Integer, (Integer, Integer)) -> Set (Integer, Integer) -> Bool
myNotMember (_,pos) s = not (member pos s)

hasNeg :: (Integer, (Integer, Integer)) -> Bool
hasNeg (_,(x,y)) = (x >= 0) && (y >= 0)

nextStates :: Integer -> (Integer, (Integer, Integer)) -> Set (Integer, Integer) 
    -> ([(Integer, (Integer, Integer))], Set (Integer, Integer))
nextStates luckyN state visited = ((nextStates' luckyN state visited)
    , union visited (fromList (map snd (nextStates' luckyN state visited))))

nextStates' :: Integer -> (Integer, (Integer, Integer)) 
    -> Set (Integer, Integer) -> [(Integer, (Integer, Integer))]
nextStates' luckyN state visited = filter (`myNotMember` visited) 
    (filter (`isOpen` luckyN) (nextStates'' state))

nextStates'' :: (Integer, (Integer, Integer)) -> [(Integer, (Integer, Integer))]
nextStates'' (move, (x,y)) = filter (hasNeg) [((move+1),((x+1),y))
    , ((move+1),(x,(y-1))), ((move+1),((x-1),y)), ((move+1),(x,(y+1)))]

solve :: (Integer, Integer) -> Integer -> [(Integer, (Integer, Integer))] 
    -> Set (Integer, Integer) -> Integer
solve _ _ [] _ = -1
solve goal luckyN ((move, pos):states) visited
    | pos == goal = move
    | otherwise = solve goal luckyN (states ++ 
        (fst (nextStates luckyN (move, pos) visited))) 
        (snd (nextStates luckyN (move, pos) visited))

solve2 :: Integer -> [(Integer, (Integer, Integer))] -> Set (Integer, Integer) 
    -> Int
solve2 _ [] _ = -1
solve2 luckyN ((move, pos):states) visited
    | move == 50 = size visited
    | otherwise = solve2 luckyN 
        (states ++ (fst (nextStates luckyN (move, pos) visited))) 
        (snd (nextStates luckyN (move, pos) visited))

main = do
    input <- readFile "d13_input.txt"
    let luckyNumber = read input :: Integer
    let answer1 = solve (31,39) luckyNumber [(0, (1, 1))] (fromList [(1, 1)])
    let answer2 = solve2 1364 [(0, (1, 1))] (fromList [(1, 1)])
    putStrLn $ "Part one: " ++ show answer1
    putStrLn $ "Part one: " ++ show answer2

File: advall-python/d13.py
from collections import deque

fileSrc = "d13_input.txt"

def getLuckyNumber(src):
    with open (src, "r") as inputfile:
        input = inputfile.read()
    return int(input)

def isWall(x, y, luckyNumber):
    s = (x*x + 3*x + 2*x*y + y + y*y + luckyNumber)
    binS = "{0:b}".format(s)
    return ((binS.count("1")) % 2) == 1

def move(state, delta, luckyNumber):
    nX = state[1][0] + delta[0]
    nY = state[1][1] + delta[1]
    nMoves = state[0] + 1
    if (nX < 0) or (nY < 0) or isWall(nX, nY, luckyNumber):
        return []
    else:
        return[(nMoves, (nX, nY))]

def solve(luckyNumber, start, goal):
    states = deque()
    startState = (0, start)
    states.append(startState)
    visited = set()
    visited.add(startState)
    while len(states) > 0:
        state = states.popleft()
        if state[1] == goal:
            return state[0]
        nextStates = move(state, (1, 0), luckyNumber) + move(state, (0, -1), luckyNumber) + move(state, (-1, 0), luckyNumber) + move(state, (0, 1), luckyNumber)
        for s in nextStates:
            if s[1] not in visited:
                states.append(s)
                visited.add(s[1])
    print "Could not find the goal!"
    return -1

def solve2(luckyNumber, start):
    states = deque()
    startState = (0, start)
    states.append(startState)
    visited = set()
    visited.add(start)
    while (len(states) > 0):
        state = states.popleft()
        if state[0] == 50:
            break
        nextStates = move(state, (1, 0), luckyNumber) + move(state, (0, -1), luckyNumber) + move(state, (-1, 0), luckyNumber) + move(state, (0, 1), luckyNumber)
        for s in nextStates:
            if s[1] not in visited:
                states.append(s)
                visited.add(s[1])
    return len(visited)

print "Part one: " + str(solve(getLuckyNumber(fileSrc), (1, 1), (31,39)))
print "Part two: " + str(solve2(getLuckyNumber(fileSrc), (1, 1)))

